src/backend/distributed/planner/intermediate_result_pruning.c

```diff
@@ -228,6 +252,22 @@ FindAllWorkerNodesUsingSubplan(HTAB *intermediateResultsHash,
 		}
 	}
 
+	/* don't include the current worker if the result will be written to local file */
+	if (entry->writeLocalFile)
+	{
+		WorkerNode *workerNode = NULL;
+		int32 localGroupId = GetLocalGroupId();
+
+		foreach_ptr(workerNode, workerNodeList)
+		{
+			if (workerNode->groupId == localGroupId)
+			{
+				workerNodeList = list_delete_ptr(workerNodeList, workerNode);
```

Comment:
Does this cause any issues given that we're iterating through this list? (not sure)

Reply:
I checked `list_delete_ptr()` implementation, and there is no harm in using it this way. The only downside is `list_delete_ptr()` iterates through the linked list to update the link on previous cell.

https://github.com/postgres/postgres/blob/master/src/backend/nodes/list.c#L795-L812

Reply:
Well, even if it doesn't cause any issues, it looks weird. Instead, I got the copy of a shallow copy of the workerNodeList

Reply:
instead of having this complicated logic to remove the local node from the list can we rewrite the loop above (the one that sets `entry->writeLocalFile` and creates `workerNodeList` from 1 loop to 2 loops?

first loop checks for and sets `entry->writeLocalFile`
second loop constructs `workerNodeList`

if performance of running the loop twice is a concern we could also fold it in 1 loop and keep track of the `workerNode` we want to remove later during the run of the first loop (nothing to remove if ptr is `NULL`).

Nevertheless the current construct is controversial and hard to follow.